src/main/java/appeng/core/worlddata/SpatialDimensionManager.java

```diff
@@ -0,0 +1,223 @@
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Vec3i;
+import net.minecraft.world.World;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.capabilities.ICapabilitySerializable;
+import net.minecraftforge.common.util.INBTSerializable;
+
+import appeng.api.storage.ISpatialDimension;
+import appeng.capabilities.Capabilities;
+
+
+public class SpatialDimensionManager implements ISpatialDimension, ICapabilitySerializable<NBTTagCompound>
+{
+	private World world;
+	private Map<Integer, StorageCellData> spatialData = new HashMap<>();
+
+	private static final int MAX_CELL_DIMENSION = 512;
+
+	public SpatialDimensionManager( World world )
+	{
+		this.world = world;
+	}
+
+	@Override
+	public World getWorld()
+	{
+		return this.world;
+	}
+
+	@Override
+	public int createNewCellDimension( Vec3i contentSize, int owner )
+	{
+		int newId = this.getNextId();
+
+		StorageCellData data = new StorageCellData();
+		data.contentDimension = contentSize;
+		data.owner = owner;
+
+		this.initCellArea( data );
+
+		this.spatialData.put( newId, data );
+
+		return newId;
+	}
+
+	@Override
+	public void deleteCellDimension( int cellStorageId )
+	{
+		// TODO cleanup blocks
+		StorageCellData removed = spatialData.remove( cellStorageId );
+		if( removed != null )
+		{
+			this.clearCellArea( removed );
+		}
+	}
+
+	@Override
+	public boolean isCellDimension( int cellStorageId )
+	{
+		return this.spatialData.containsKey( cellStorageId );
+	}
+
+	@Override
+	public int getCellDimensionOwner( int cellStorageId )
+	{
+		StorageCellData cell = this.spatialData.get( cellStorageId );
+		if( cell != null )
+		{
+			return cell.owner;
+		}
+		return -1;
+	}
+
+	@Override
+	public BlockPos getCellDimensionOrigin( int cellStorageId )
+	{
+		if( this.isCellDimension( cellStorageId ) )
+		{
+			return getBlockPosFromId( cellStorageId );
+		}
+		return null;
+	}
+
+	@Override
+	public Vec3i getCellContentSize( int cellStorageId )
+	{
+		StorageCellData cell = this.spatialData.get( cellStorageId );
+		if( cell != null )
+		{
+			return cell.contentDimension;
+		}
+		return null;
+	}
+
+	@Override
+	public boolean hasCapability( Capability<?> capability, EnumFacing facing )
+	{
+		return capability == Capabilities.SPATIAL_DIMENSION;
+	}
+
+	@Override
+	public <T> T getCapability( Capability<T> capability, EnumFacing facing )
+	{
+		if( capability == Capabilities.SPATIAL_DIMENSION )
+		{
+			return (T) this;
+		}
+		return null;
+	}
+
+	@Override
+	public NBTTagCompound serializeNBT()
+	{
+		final NBTTagCompound ret = new NBTTagCompound();
+		final NBTTagList list = new NBTTagList();
+
+		for( Map.Entry<Integer, StorageCellData> entry : this.spatialData.entrySet() )
+		{
+			final NBTTagCompound nbt = entry.getValue().serializeNBT();
+			nbt.setInteger( "id", entry.getKey() );
+			list.appendTag( nbt );
+		}
+		ret.setTag( "spatial_data", list );
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "spatial_data" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1192)